### PR TITLE
Automate creation of new builders with @build

### PR DESCRIPTION
**Pull request by jbrea**
When prototyping, I would prefer not to have to write explicitly a new builder for any architecture I want to try.

### CODE REVIEW

**Comment by ablaom** [comment]
Something like this sound like a good idea. 

I have a few suggestions:

1. Unlike all the other examples of builders in the docs, this one simply ignores the shape of the data, and will only work if the shape of the data matches the architecture as specified.  That is, it does not "dynamically" define a Flux model, which is the main idea of a "builder". I expect this will create cognitive dissonance and cause confusion. This would probably be mitigated by a different name, such as `static_builder`.

2. I suggest you don't use a macro here, as you can use a parametric type instead. This will be more robust, work in any scope, avoid issues with non-reproducible type names (eg, serialization). Something like:

```julia
StaticBuilder{C} <: MLJFlux.Builder
    chain::C
end

MLJFlux.build(b::StaticBuilder, rng, n_in, n_out, shape...) = b.chain
```
Then user can do, eg:

```
builder = StaticBuilder(Chain(Dense(3, 4), Dense(4,1))
```

What do you think?

**Comment by jbrea** [comment]
Thanks for your feedback!

1. This can be mitigated with variables `n_in`, `n_out`, `n_channels` that can be used in the macro. See updated code.
2. I guess the parametric type approach would lead to some quite unexpected behaviour, because the network gets built only once. For example, running `evaluate!` with cross-validation would always use the same network, wouldn't it? But you are right, the macro approach doesn't play well with serialization. I don't know how to solve this, except with a note in the documentation that this  macro is meant for prototyping and a custom, named builder should be used for production.

**Comment by ablaom** [comment]
> I guess the parametric type approach would lead to some quite unexpected behaviour, because the network gets built only once.

You are right! Here is a variation on my original idea, now with a macro which I think is more robust than the original. I'm still ignoring the shape and rng here. 

```julia
struct GenericBuilder{F} <: MLJFlux.Builder
    apply::F # function with no arguments
end

macro builder(ex)
    quote
        GenericBuilder(() -> $ex)
    end
end

MLJFlux.build(b::GenericBuilder, args...) = b.apply()

julia> b = @builder Dense(2,3)
GenericBuilder{var"#9#10"}(var"#9#10"())

julia> MLJFlux.build(b, 123, 2, 3).weight
3×2 Matrix{Float32}:
 0.518701  0.581197
 0.283643  0.356529
 0.613235  0.709665

julia> MLJFlux.build(b, 123, 2, 3).weight
]3×2 Matrix{Float32}:
  0.981433  -0.29414
  0.759679   0.248205
 -0.343442  -0.0562241

```

What do you think?

I've not looked over your other changes yet.

**Comment by jbrea** [comment]
Ah, yes, thanks, that's the way to go! I've implemented this approach in the latest commit.

File: README.md [hunk 1]
```diff
@@ -275,6 +275,9 @@ MLJ provides two simple builders out of the box. In all cases weights
 
 See Table 1 above to see how `n_in` and `n_out` relate to the data.
 
+Alternatively, `MLJFlux.@builder(neural_net)` automatically creates a builder
+for any valid `neural_net`, e.g. `MLJFlux.@builder(Dense(n_in, n_out, tanh))`.
+
```
Comment (ablaom) [inline README.md:280]: The doc string you wrote is great, but this addition to the README could be a little a little more detailed. Maybe something along the lines of 

"Alternatively, `use MLJFlux.@builder(neural_net)` to automatically create a builder
 for any valid Flux chain expression `neural_net`, where the symbols `n_in`, `n_out`, `n_channels` and `rng` can appear literally, with the interpretations explained above. For example,

    builder = MLJFlux.@builder Chain(Dense(n_in, 128), Dense(128, n_out, tanh))
"

**Review by ablaom — COMMENTED** [review]
This is a really neat addition, thanks!

File: README.md (hunk 1 above)
Comment (jbrea) [reply README.md:286]: Sounds good! Thanks.

**Pull request merged**